Add --bail option to fail upon first error

stanhu opened the pull request:
When a failure happens in the build step, it's often difficult to debug
with `--verbose` on because the error may have occurred in the middle of
the output.  This option now fails upon first error and shows the error
message to make it easier to diagnose and fix problems.

Closes #1924

tdreyno commented:
Seems smart. Might be better to just `exit(1)` rather that raising, which will make even more noise.

tdreyno commented:
If you could update the `CHANGELOG` as well, that'd be great.

stanhu commented:
I used `exit(1)` first, but then I got:

```
Parallel::DeadWorker: Parallel::DeadWorker
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:74:in `rescue in work'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:71:in `work'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:382:in `block (4 levels) in work_in_processes'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:495:in `with_instrumentation'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:381:in `block (3 levels) in work_in_processes'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:369:in `loop'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:369:in `block (2 levels) in work_in_processes'
  /Users/stanhu/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/parallel-1.12.1/lib/parallel.rb:206:in `block (2 levels) in in_threads'
```

I changed to the exception because it looked like that was captured and handled, but I'm happy to make the change if you think we're better off with the `exit(1)` and disabling parallel works.

Added a CHANGELOG entry.

tdreyno commented:
Ah, that makes sense.

stanhu commented:
@tdreyno Would you mind tagging a new release with this? Thanks!

tdreyno commented:
For 4.x or 5.x beta?

stanhu commented:
4.x.

tdreyno commented:
I'll get one out on Tuesday

tdreyno commented:
@stanhu Done. 4.3.3

stanhu commented:
@tdreyno Thank you! This will help our team troubleshoot failures: https://gitlab.com/gitlab-com/www-gitlab-com/merge_requests/19336
